book/src/rfc/const_generic.md

```diff
@@ -0,0 +1,265 @@
+
+Enabling const generic will allow define array sizes with constant variables. This would enhance function reusability and significantly improve modularization.
+
+## Detailed design
+
+### Declare a const argument
+Enable functions to define a constant argument, which can be used to define arrays with a default element or as an array size of the return type:
+
+```rust
+fn const_generic(const cst: Field) -> [Field; cst] {
+    let mut xx = [1; cst];
+    for ii in 1..cst {
+        xx[ii] = 2;
+    }
+    return xx;
+}
+```
+
+### Applying a const as a argument
+Use an existing constant to set array size directly in function parameters:
+
+```rust
+const cst = 300;
+fn const_generic(yy: Field) -> [Field; cst] {
+    let xx = [1; cst];
+    return xx;
+}
+```
+
+### Builtin functions
+
+Allow builtin functions to be defined with const arguments:
+
+```rust
+use std::to_bits;
+use std::from_bits;
+
+const num_bits = 8;
+
+fn main() {
+    let val1 = 101;
+    let bits = to_bits(num_bits, val); // bits is array ComputedExpr with len
+    let val2 = from_bits(bits); // from_bits builtin impl knows the size of the array
+    assert_eq!(val1, val2);
+}
+```
+
+The `to_bits` builtin function accepts a const variable to determine the size of the return array.
+
+```rust
+fn to_bits(const num_bits: Field, val: Field) -> [Field; num_bits]
+```
+
+The from_bits builtin function can determine the number of bits to convert based on the information that the `ComputedExprKind::Array` embeds in the `bits` variable. We will explain on `ComputedExpr` at the end to show how to improve the structure of the inputs and outputs of these builtin functions.
+
+## Type checking const variable
+
+The existing type checker is designed to do local type checks. For the cases of const variable as array size, the type checker has to trace the values globally. This is out of the scope for the design of the current type checker.
+
+Instead of checking the const variable in the type checker, we can check it in the circuit generation phase. 
+
+### Review how VarOrRef works
+ 
+`VarOrRef` is responsible for containing the underlying variables to pass around during circuit generation.
+
+For example:
+```rust
+houses[1].rooms[2].size
+```
+
+The variable `houses` holds all the `cvars` to represent the underlying low level variables. If `houses` is a mutable variable, it will be a reference, which records the following instead of holding the actual `cvars` array:
+ - the variable name `houses` in local scope
+ - start index of the `cvars` to represent the current access path
+ - the length of the `cvars` to represent size of the element targeted by the current access path
+
+So when it wants to update the elements inside the `houses`, it will need to reference the start index and the length of the `cvars` of the elements, so that it can narrow down the scope of the `cvars` to be updated.
```

Comment:
ah OK I think I understand why I initially did this, because when you do 

```rust
houses[1].rooms[2] = 5;
```

you end up parsing expression after expression recursively, and as you go from left to right you lose information. It sorts of look like this AST:

```
Assignment(ArrayAccess(FieldAccess(ArrayAccess("houses", 1), "rooms"), 2), 5)
```

and so you want to be able to pinpoint exactly what exactly is being assigned something (so that it can be reflected later in the scope where `houses` exist)

Reply:
Exactly!